### DISCUSSION TITLE
Model conversion to CV25

### DISCUSSION THREAD

**Discussion by LaurentAIA** [AxisCommunications/discussions#604] [ACAP] [answered]
Hello,

I am trying to convert my model to CV25.
I have a .tflite model working well on artpec8 cameras and now I would like to convert to a model compatible with CV25 (.bin) : 

<img width="492" alt="Screenshot 2024-04-16 at 11 44 16" src="https://github.com/AxisCommunications/discussions/assets/113306968/4218a1d6-412b-4fd7-9b7e-396bdcb63477">

so I tried to follow this example : [tensorflow-to-larod-cv25#compiling-for-cv25-dlpu](https://github.com/AxisCommunications/acap-native-sdk-examples/tree/main/tensorflow-to-larod-cv25#compiling-for-cv25-dlpu)

Got the Ambarella Toolchain and installed it on a Ubuntu20.04 with Python 3.8

And now I am having difficulties during the conversion process :

After creating the image list, I go to the tfparser.py step but the vas file is not created :

➜  LaurentAIA tfparser.py -p laurent.tflite \                                                                                                               
-o fmodel_convert \
-of ./out_model \
-isrc "is:1,3,320,320|iq|i:input_1=./dra_image_bin/dra_bin_list.txt|ic:1|idf:0,0,0,0" \
-odst "o:TFLite_Detection_PostProcess|odf:fp32" \
-odst "o:TFLite_Detection_PostPtocess:1|odf:fp32" \
-odst "o:TFLite_Detection_PostProcess:2|odf:fp32" \
-odst "o:TFLite_Detection_PostProcess:3|odf:fp32" \
-c act-force-fx8,coedd-force-fx8 

2024-04-16 15:19:41.863821: I tensorflow/stream_executor/platform/default/dso_loader.cc:50] Successfully opened dynamic library libcudart.so.12
WARNING:tensorflow:Deprecation warnings have been disabled. Set TF_ENABLE_DEPRECATION_WARNINGS=1 to re-enable them.
WARNING:root:Limited tf.compat.v2.summary API due to missing TensorBoard installation.
Warning [parser]: removing existing dir /home/bulltech/code/LaurentAIA/out_model
WARNING:tensorflow:From /usr/local/amba-cv-tools-basic.2.5.6.0.1245.ubuntu-20.04/cv2/tv2/release/frameworklibs/cv2.basic.2.5.6.0.1245.ubu2004/lib/python3.8/site-packages/frameworklibs/tensorflow/tflite/tf_lite_supported_ops.py:60: The name tf.nn.max_pool is deprecated. Please use tf.nn.max_pool2d instead.

WARNING:tensorflow:From /usr/local/amba-cv-tools-basic.2.5.6.0.1245.ubuntu-20.04/cv2/tv2/release/frameworklibs/cv2.basic.2.5.6.0.1245.ubu2004/lib/python3.8/site-packages/frameworklibs/tensorflow/tflite/tf_lite_supported_ops.py:60: The name tf.nn.max_pool is deprecated. Please use tf.nn.max_pool2d instead.

WARNING:tensorflow:From /usr/local/amba-cv-tools-basic.2.5.6.0.1245.ubuntu-20.04/cv2/tv2/release/frameworklibs/cv2.basic.2.5.6.0.1245.ubu2004/lib/python3.8/site-packages/frameworklibs/tensorflow/tflite/tf_lite_supported_ops.py:61: The name tf.nn.avg_pool is deprecated. Please use tf.nn.avg_pool2d instead.

WARNING:tensorflow:From /usr/local/amba-cv-tools-basic.2.5.6.0.1245.ubuntu-20.04/cv2/tv2/release/frameworklibs/cv2.basic.2.5.6.0.1245.ubu2004/lib/python3.8/site-packages/frameworklibs/tensorflow/tflite/tf_lite_supported_ops.py:61: The name tf.nn.avg_pool is deprecated. Please use tf.nn.avg_pool2d instead.

WARNING:tensorflow:From /usr/local/amba-cv-tools-basic.2.5.6.0.1245.ubuntu-20.04/cv2/tv2/release/frameworklibs/cv2.basic.2.5.6.0.1245.ubu2004/lib/python3.8/site-packages/frameworklibs/tensorflow/tflite/tf_lite_supported_ops.py:104: The name tf.batch_to_space_nd is deprecated. Please use tf.batch_to_space instead.

WARNING:tensorflow:From /usr/local/amba-cv-tools-basic.2.5.6.0.1245.ubuntu-20.04/cv2/tv2/release/frameworklibs/cv2.basic.2.5.6.0.1245.ubu2004/lib/python3.8/site-packages/frameworklibs/tensorflow/tflite/tf_lite_supported_ops.py:104: The name tf.batch_to_space_nd is deprecated. Please use tf.batch_to_space instead.

WARNING:tensorflow:From /usr/local/amba-cv-tools-basic.2.5.6.0.1245.ubuntu-20.04/cv2/tv2/release/frameworklibs/cv2.basic.2.5.6.0.1245.ubu2004/lib/python3.8/site-packages/frameworklibs/tensorflow/tflite/tf_lite_supported_ops.py:105: The name tf.space_to_batch_nd is deprecated. Please use tf.space_to_batch instead.

WARNING:tensorflow:From /usr/local/amba-cv-tools-basic.2.5.6.0.1245.ubuntu-20.04/cv2/tv2/release/frameworklibs/cv2.basic.2.5.6.0.1245.ubu2004/lib/python3.8/site-packages/frameworklibs/tensorflow/tflite/tf_lite_supported_ops.py:105: The name tf.space_to_batch_nd is deprecated. Please use tf.space_to_batch instead.

INFO: 04/16/2024 15:19:42.911097 logger.py:372 [TFLite_Parser] Application logs will be written to the log directory: ./logs
INFO: 04/16/2024 15:19:42.950805 tf_lite_v2.py:254 [TFLite_Parser] [MAIN] TFLite model provided by the user: /home/bulltech/code/LaurentAIA/laurent.tflite
INFO: 04/16/2024 15:19:42.950999 tf_lite_v2.py:255 [TFLite_Parser] [MAIN] List of user output nodes: ['TFLite_Detection_PostProcess', 'TFLite_Detection_PostPtocess:1', 'TFLite_Detection_PostProcess:2', 'TFLite_Detection_PostProcess:3']
INFO: 04/16/2024 15:19:42.951050 tf_lite_v2.py:256 [TFLite_Parser] [MAIN] Do not use quantization data formats: False
INFO: 04/16/2024 15:19:42.951091 tf_lite_v2.py:257 [TFLite_Parser] [MAIN] Invoke Asymmetric adjustement: False
INFO: 04/16/2024 15:19:42.951131 tf_lite_v2.py:258 [TFLite_Parser] [MAIN] Schema file path: /usr/local/amba-cv-tools-basic.2.5.6.0.1245.ubuntu-20.04/cv2/tv2/release/frameworklibs/cv2.basic.2.5.6.0.1245.ubu2004/bin/schema.fbs
INFO: 04/16/2024 15:19:42.951169 tf_lite_v2.py:259 [TFLite_Parser] [MAIN] FLATC executable path: /usr/local/amba-cv-tools-basic.2.5.6.0.1245.ubuntu-20.04/cv2/tv2/release/frameworklibs/cv2.basic.2.5.6.0.1245.ubu2004/bin/flatc
INFO: 04/16/2024 15:19:42.951207 tf_lite_v2.py:260 [TFLite_Parser] [MAIN] Checkpoint path: /home/bulltech/code/LaurentAIA//reconstructed_pb_model/laurent_ckpt
INFO: 04/16/2024 15:19:42.951244 tf_lite_v2.py:261 [TFLite_Parser] [MAIN] Output protobuf reconstructed: /home/bulltech/code/LaurentAIA//reconstructed_pb_model/laurent_reconstructed.pb
INFO: 04/16/2024 15:19:42.951290 tf_lite_v2.py:285 [TFLite_Parser] [FLATC EXE] Converting tflite model to JSON file using flatc executable
INFO: 04/16/2024 15:19:44.369847 tf_lite_v2.py:427 [TFLite_Parser] [RECONSTRUCT PROTO] Convert JSON file to protobuf
INFO: 04/16/2024 15:19:44.716016 tf_lite_v2.py:316 [TFLite_Parser] [SET IO NODE] Input of the model: ['normalized_input_image_tensor']
INFO: 04/16/2024 15:19:44.716168 tf_lite_v2.py:317 [TFLite_Parser] [SET IO NODE] Output of the model: ['TFLite_Detection_PostProcess', 'TFLite_Detection_PostPtocess:1', 'TFLite_Detection_PostProcess:2', 'TFLite_Detection_PostProcess:3']

**************************************************

WARNING!!! CUSTOM OP PRESENT!

PRIMARY INPUTS: 

	normalized_input_image_tensor  [1, 320, 320, 3]

PRIMARY OUTPUTS: 

	TFLite_Detection_PostProcess    []
	TFLite_Detection_PostProcess:1  []
	TFLite_Detection_PostProcess:2  []
	TFLite_Detection_PostProcess:3  []

CUSTOM OP NAME: TFLite_Detection_PostProcess

	CUSTOM OP INPUTS: 

		Squeeze         [1, 2034, 4]
		convert_scores  [1, 2034, 2]
		anchors         [2034, 4]

	CUSTOM OP OUTPUTS: 

		TFLite_Detection_PostProcess  []

Recommendation: 
Use GraphSurgery tool to cut the graph into separate protobufs between the custom ops.
The individual protobufs may still have unsupported ops. Please run tf_print_graph_summary & GraphSurgery on both of them independently to make sure the Graph is VP friendly.

**************************************************

**Comment by github-actions** [comment]
*This automatically generated reply acts as a friendly reminder.*

*Answers to your questions will most often come from the community, from developers like yourself. You will, from time to time, find that Axis employees answers some of the questions, but this is not a guarantee. Think of the discussion forum as a complement to other support channels, not a replacement to any of them.*

*If your question remains unanswered for a period of time, please revisit it to see whether it can be improved by following the guidelines listed in [Axis support guidelines](https://github.com/orgs/AxisCommunications/discussions/13).*

**Comment by Corallo** [comment]
Hi @LaurentAIA 

On a Cv25 device, it is not possible to run custom ops like the post-processing one that you have in ssd-mobilenet.

With the ambarella toolkit, there should be a  "graph_surgery.py" script that can be used to cut out the network before the custom OP
Something like this:

```
graph_surgery.py tf -p ./ssd_mobilenet_v2_coco_quant_postprocess.tflite -o frozen_graph_def_after_surgery.pb \
-isrc "i:normalized_input_image_tensor|is:1,300,300,3" \
-on Squeeze2,convert_scores1 \
-t ConstantifyShapes
```

After the surgery, you can proceed with your model conversion.

Note that without that post-processing part, you will have to do the post-processing by yourself in your C code:
That's what we do in our [object-detection-cv25 example](https://github.com/AxisCommunications/acap-native-sdk-examples/blob/main/object-detection-cv25/app/postprocessing.c)
In alternative, you can also split the post-processing, and run it on CPU using the larod api, unfortunately, we don't have an example for that.

**Reply by LaurentAIA** [reply]
Thank you @Corallo, I will have a look and try to perform surgery on my model, let's hope it will survive it :))

Will keep you updated.

**Reply by LaurentAIA** [reply, marked as answer]
Hello @Corallo, I managed to get the model converted by cutting 2 outputs. I will post later the process I went through.

Edit : 

I performed the surgery with that command : 

`graph_surgery.py tf -p ./model.tflite -o frozen_graph_def_after_surgery.pb -isrc "i:normalized_input_image_tensor|is:1,320,320,3" -on Squeeze,convert_scores -t ConstantifyShapes`

Then I prepared the dataset for quantization.

And then comes the parsing step, so I ran : 

```
tfparser.py \
      -p frozen_graph_def_after_surgery.pb \
      -isrc "i:normalized_input_image_tensor=dra_image_bin/dra_bin_list.txt|idf:0,0,0,0|is:1,3,320,320|iq|im:127.5|ic:127.5" \
      -odst "o:convert_scores|ot:0,2,3,1|os:1,1,1,4068|odf:fp32" \
      -odst "o:Squeeze|ot:0,2,3,1|os:1,1,1,8136|odf:fp32" \
      -o converted_model \
      -of out_tf_parser
```

However that last command is showing some progress first until it reaches 100%, and then it continues to run indefinitely ... Not sure what is wrong.

<img width="1602" alt="Screenshot 2024-07-10 at 16 29 32" src="https://github.com/AxisCommunications/discussions/assets/113306968/9b7d7d72-57f3-47ed-ab15-0891601a2ddc">

**Comment by NithinVoxela** [comment]
@LaurentAIA how did you get the toolchain from Ambarella

**Reply by vivekatoffice** [reply]
Hi @NithinVoxela ,
This may help:

<img width="699" height="451" alt="image" src="https://github.com/user-attachments/assets/7eae3de2-e650-44ff-b7fc-b9f49b420aba" />

[Image Reference](https://developer.axis.com/computer-vision/computer-vision-on-device/dlpu-model-conversion/#cv25)

**Reply by NithinVoxela** [reply]
Hi @vivekatoffice i dont have an account in Ambarella also there no option to register for new customer in provided link, can you please tell me how to contact them to get toolchain? Thanks in advance

**Reply by vivekatoffice** [reply]
Hi @NithinVoxela ,
You may try these contacts: [www.ambarella.com/contact-us](https://www.ambarella.com/contact-us/). The agreement with your company and Ambarella may be required.

I can confirm that, when logged in with my company email account, I have access to download the following files:
[customer.ambarella.com/ng/cvtools](https://customer.ambarella.com/ng/cvtools)

<img width="1911" height="1092" alt="image" src="https://github.com/user-attachments/assets/89077541-8f74-4d56-8f02-7841e53d98c1" />

**Reply by NithinVoxela** [reply]
ok i will contact them, Thank you!

**Reply by NithinVoxela** [reply]
Hi @vivekatoffice , we contacted Ambarella they're telling to get it from AXIS and AXIS people are telling it to get it from Ambarella

**Reply by vivekatoffice** [reply]
Hi @NithinVoxela , I suggest creating a new discussion [here](https://github.com/orgs/AxisCommunications/discussions). As this discussion is already marked **Answered**, less visibility to other [Axis Developer Community](https://www.axis.com/developer-community) members.

**Reply by vivekatoffice** [reply]
Hi @NithinVoxela , 
Do you have this **resister** link when you Visit [this](https://customer.ambarella.com/ng/login?refid=EAAAALA9%2fIfpWDCn53oQJDd5FKfzsrI0fWXYseVTpgnkJHV1) web page to access the tools.

<img width="560" height="667" alt="image" src="https://github.com/user-attachments/assets/72fbf35a-8996-4055-b6a7-b5800eff3e36" />

> Note: use the complete link: https://customer.ambarella.com/ng/login?refid=EAAAALA9%2fIfpWDCn53oQJDd5FKfzsrI0fWXYseVTpgnkJHV1

**Reply by NithinVoxela** [reply]
yes i see register button

**Reply by vivekatoffice** [reply]
Once you register you can get access to the toolchain. According to my colleague:
_"register" is needed first and then you can get the toolchain. Version 2.5.6 works well. the documents are also included._

Hope this will help 😉

**Reply by NithinVoxela** [reply]
ok, Thank you Vivek for the help
